addons/mail/tests/test_thread_controller.py

```diff
@@ -0,0 +1,196 @@
+
+    def _execute_subtests(self, record, subtests):
+        for user, allowed, *args in subtests:
+            route_kw = {
+                "context": {"mail_create_nosubscribe": True, "mail_post_autofollow": False},
+                **(args[0] if args else {}),
+            }
+            post_data = {
+                "body": self.content,
+                "message_type": "comment",
+                "subtype_xmlid": "mail.mt_comment",
+                **(args[1] if len(args) > 1 else {}),
+            }
+            kwargs = args[2] if len(args) > 2 else {}
+            self._authenticate_user(user)
+            with self.subTest(record=record, login=user.login, route_kw=route_kw):
+                if allowed:
+                    self._message_post(record, post_data, route_kw)
+                    message = self._find_message(record)
+                    self.assertTrue(message)
+                    expected_author = kwargs.get("partner")
+                    if user == self.public_w_guest and not expected_author:
+                        self.assertEqual(message.author_guest_id, self.guest)
+                    else:
+                        self.assertEqual(message.author_id, expected_author or user.partner_id)
+                    if "partner_ids" in post_data:
+                        self.assertEqual(
+                            set(message.partner_ids.ids),
+                            set(kwargs.get("expected_partner_ids", post_data["partner_ids"])),
+                        )
+                    if "partner_emails" in route_kw:
+                        self.assertEqual(
+                            set(message.partner_ids.mapped("email")),
+                            set(kwargs.get("expected_partner_emails", route_kw["partner_emails"])),
+                        )
+                else:
+                    with self.assertRaises(JsonRpcException, msg="werkzeug.exceptions.NotFound"):
+                        self._message_post(record, post_data, route_kw)
+
+    def _message_post(self, record, post_data, route_kw):
+        self.make_jsonrpc_request(
+            route=f"{self.env.user.get_base_url()}/mail/message/post",
+            params={
+                "thread_model": record._name,
+                "thread_id": record.id,
+                "post_data": post_data,
+                **route_kw,
+            },
+        )
+
+    def _find_message(self, record):
+        return self.env["mail.message"].search(
+            [("res_id", "=", record.id), ("model", "=", record._name)], order="id desc", limit=1
+        )
+
+
+@odoo.tests.tagged("-at_install", "post_install")
+class TestThreadController(TestThreadControllerCommon):
+    def test_partner_message_post_access(self):
+        """Test access of message_post on partner record."""
+        record = self.env["res.partner"].create({"name": "Test Partner"})
+        self._execute_subtests(
+            record,
+            (
+                (self.no_user, False),
```

Comment:
Tuples in access tests are becoming common. It would be much better to either use dict or namedtuple for readability/maintainability.

Reply:
The problem with dict is that they are too verbose, I believe there is value to keep these lines all single-line and not too long.

Aren't `namedtuple` useful only when accessing the data with key? But when accessing the data it's rather obvious which is which as they are stored in dedicated variable.

Reply:
Small tuples like this one are fine, but when they grow in value, it's not obvious what is what.
Below this test, there are 4 entries tuples and I have no idea what the third value is. I will have to read  `self._execute_subtests` to know that.

We could also try to document or type them somehow?